dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/data/DefaultAnalyticsService.java

```diff
@@ -1364,9 +1372,86 @@ private void addRawData( DataQueryParams params, Grid grid )
             .withOutputFormat( OutputFormat.ANALYTICS )
             .withSkipMeta( true ).build();
 
+        // each row contains: dimension uid | period | value
         Grid grid = getAggregatedDataValueGridInternal( dataSourceParams );
 
-        return grid.getAsMap( grid.getWidth() - 1, DimensionalObject.DIMENSION_SEP );
+        MultiValuedMap<String, DimensionItemWithValue> result = new ArrayListValuedHashMap<>();
+
+        final List<DimensionalItemObject> basePeriods = params.getPeriods();
+
+        for ( List<Object> row : grid.getRows() )
+        {
+            if ( isInBasePeriod( basePeriods, (String) row.get( 1 ) ) )
+            {
+                // Key is composed of [uid-period]
+                final String key = StringUtils.join(
+                    ArrayUtils.remove( row.toArray( new Object[0] ), grid.getWidth() - 1 ),
+                    DimensionalObject.DIMENSION_SEP );
+
+                final DimensionalItemObject dimensionalItemObject = findDimensionalItem( (String) row.get( 0 ), items ).get( 0 );
+                DimensionalItemObject clone = dimensionalItemObject;
+                if ( dimensionalItemObject.getPeriodOffset() != 0 )
+                {
+                    List<Object> periodOffsetRow = getPeriodOffsetRow( grid, dimensionalItemObject, (String)row.get(1), dimensionalItemObject.getPeriodOffset());
+                    if ( periodOffsetRow != null )
+                    {
+                        result.put( key,
+                            new DimensionItemWithValue( dimensionalItemObject,
+                                (Double) periodOffsetRow.get( grid.getWidth() - 1 ) ) );
+
+                    } // TODO throw exception?
+                    clone = SerializationUtils.clone( dimensionalItemObject );
+                    clone.resetPeriodOffset();
+                }
+
+                result.put( key,
+                    new DimensionItemWithValue( clone, (Double) row.get( grid.getWidth() - 1 ) ) );
+            }
+        }
+        
+        return result;
+    }
+    
+    private List<Object> getPeriodOffsetRow( Grid grid, DimensionalItemObject dimItem, String isoPeriod, int offset )
```

Comment:
Could we move this method to a utility class, add Javadoc and add unit test? E.g.`AnayticsUtils`.

Reply:
Fix

Reply:
Fix